### PR TITLE
Asyncio support for Scrapy | Gsoc 2018

### PR DESCRIPTION
**Pull request by yashrsharma44**
I have made the initial pull request, I am working on the requirements as we discussed, and I will update this PR, as things are completed @dangra .

### CODE REVIEW

**Comment by yashrsharma44** [comment]
Hey @dangra, I have made the following changes - 
1. Cleaning of the code base, removing any unnecessary comments
2. Allowing Fetch to accept a request parameter.
3. Enabled the request to be passed through spider middlewares( though Fetch is an awaitable, so it cannot be passed through spidermiddlewares)
4. Sharing the crawler and spider object implicitly.

File: scrapy/utils/ossignal.py [hunk 1]
```diff
@@ -2,6 +2,14 @@
...
 import signal
 
+import asyncio
+'''
+from twisted.internet import asyncioreactor
+try:
+    asyncioreactor.install(asyncio.get_event_loop())
+except Exception:
+    pass
+'''
```
Comment (dangra) [inline scrapy/utils/ossignal.py]: what is the point of importing `asyncio` here?

File: scrapy/Fetch/__init__.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+from scrapy.core import engine
+
+from scrapy.utils.log import logformatter_adapter, failure_to_exc_info
+from scrapy import signals
+from scrapy.core.downloader import Downloader 
+from twisted.internet.defer import Deferred
+import asyncio
+import logging
+import scrapy
+import warnings
```
Comment (dangra) [inline scrapy/Fetch/__init__.py]: See https://www.python.org/dev/peps/pep-0008/#imports for import standard rules.

Specially ordering:

    1. Standard library imports.
    2. Related third party imports.
    3. Local application/library specific imports.

File: scrapy/core/downloader/__init__.py [hunk 3]
```diff
@@ -6,7 +6,16 @@
...
 import six
-from twisted.internet import reactor, defer, task
+import asyncio
+'''
+from twisted.internet import asyncioreactor
+try:
+    asyncioreactor.install(asyncio.get_event_loop())
+except Exception:
+    pass
+'''
```
Comment (dangra) [inline scrapy/core/downloader/__init__.py]: why are you importing `asyncio` in many different places? If it needs to be imported before anything else please put it on `scrapy/__init__.py` only.

**Comment by yashrsharma44** [comment]
Hey @kmike, I was looking to get this code merged, could you provide updates, if the code is suitable/unsuitable to be merged, leaving aside tests and documentation? If it is suitable, I would gladly complete the tests and documentation.

File: scrapy/Fetch/__init__.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+from scrapy import signals
+from scrapy.core.downloader import Downloader 
+
+
+
+logger = logging.getLogger(__name__)
+
+
+
+class Fetch():
```
Comment (kmike) [inline scrapy/Fetch/__init__.py]: I think we should make Request class to be awaitable, not create a separate Fetch class.

File: scrapy/Fetch/__init__.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    def __init__(self, Request):
+        self.Request = Request
+        self.crawler = engine.concraw
+        self.spider = engine.conspid
```
Comment (kmike) [inline scrapy/Fetch/__init__.py]: There can be several Crawlers running at the same time, and several spiders; we can't use global variables. I think to access Crawler contextvars should be used.

File: scrapy/__main__.py [hunk 6]
```diff
@@ -1,3 +1,4 @@
+import asyncio
```
Comment (kmike) [inline scrapy/__main__.py]: why is this import needed?

File: scrapy/core/downloader/handlers/http11.py [hunk 7]
```diff
@@ -8,6 +8,7 @@
 from six.moves.urllib.parse import urldefrag
 
 from zope.interface import implementer
+from twisted.internet import reactor
```
Comment (kmike) [inline scrapy/core/downloader/handlers/http11.py]: why is this import needed?

File: scrapy/core/engine.py [hunk 8]
```diff
@@ -4,29 +4,39 @@
...
 from scrapy.http import Response, Request
 from scrapy.utils.misc import load_object
+from scrapy.utils.defer import mustbe_deferred
 from scrapy.utils.reactor import CallLaterOnce
 from scrapy.utils.log import logformatter_adapter, failure_to_exc_info
 
 logger = logging.getLogger(__name__)
 
+conspid
+concraw
```
Comment (kmike) [inline scrapy/core/engine.py]: stray lines?

File: scrapy/spidermiddlewares/depth.py [hunk 9]
```diff
@@ -54,5 +56,11 @@ def _filter(request):
             response.meta['depth'] = 0
             if self.verbose_stats:
                 self.stats.inc_value('request_depth_count/0', spider=spider)
-
-        return (r for r in result or () if _filter(r))
+        result = list_to_simplevalue(result)
+        if isinstance(result, GeneratorType):
```
Comment (kmike) [inline scrapy/spidermiddlewares/depth.py]: It'd be much better if the feature doesn't require modifications in existing middlewares; there must be a way to support it in a framework. Otherwise users will have to modify all their custom middlewares in a similar way.

**Comment by kmike** [comment]
Hey @yashrsharma44! Unfortunately, I think the current code is very far from being merged :( There can be some useful parts - maybe some of the modifications to the `inline_requests` library, or helpers like `asyncfut_parallel` - but it is very hard for me to assess, because other code needs a rewrite, and it is not clear if these helpers will be useful after a rewrite or not.

File: scrapy/Fetch/__init__.py (hunk 4 above)
Comment (yashrsharma44) [reply scrapy/Fetch/__init__.py]: Regarding awaiting Request class, I was having circular import issues.

**Comment by kneufeld** [comment]
I'm repeating a comment I left on #1455 in case it's more relevant here:

My friend @meejah wrote https://github.com/crossbario/txaio which is `a helper library for writing code that runs unmodified on both Twisted and asyncio`

This may help bridge the twisted/asyncio/python2/python3 gaps?

**Comment by nctl144** [comment]
Hi @kmike, do you think I should make some more progress on this PR?

**Comment by yashrsharma44** [comment]
My 2cents. I was stuck with using contextvars, so that can be looked into.

File: scrapy/crawler.py [hunk 10]
```diff
@@ -2,11 +2,25 @@
 import signal
 import logging
 import warnings
+try:
+    import asyncio
+except ImportError:
+    warnings.warn("Please install asyncio to support asyncio frameworks. Without asyncio, frameworks might not be supported!")
```
[the remark this answers is not in the document]
Comment (nctl144) [reply scrapy/crawler.py]: @kmike I suppose we should change it to "this project is using Python < 3.4" right?

**Comment by wRAR** [comment]
This is no longer needed but was useful, thanks.
